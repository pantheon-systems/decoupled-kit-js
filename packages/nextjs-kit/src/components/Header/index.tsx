import {
	Header as ReactKitHeader,
	isNavItemArray,
	type NavHeaderProps,
	type NavItem,
} from '@pantheon-systems/react-kit/components/Header';
import Link from 'next/link';
import { useRouter } from 'next/router';
import { useEffect, useState } from 'react';

const Header = (props: NavHeaderProps) => {
	const { events } = useRouter();
	const [isOpen, setIsOpen] = useState(false);
	const handleOpen = () => setIsOpen((prev) => !prev);
	useEffect(() => {
		events.on('routeChangeComplete', () => {
			setIsOpen(false);
		});

		return () => {
			events.off('routeChangeComplete', () => {
				setIsOpen(false);
			});
		};
	});
	return (
		<ReactKitHeader
			{...props}
			linkComponent={props?.linkComponent ? props.linkComponent : Link}
			mobileNavHandler={[isOpen, handleOpen]}
		/>
	);
};

export { Header, NavHeaderProps, NavItem, isNavItemArray };
