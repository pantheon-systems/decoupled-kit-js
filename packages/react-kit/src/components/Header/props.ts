import { Props as FocusTrapProps } from 'focus-trap-react';
/**
 * A navigation item is a tuple of a label and a href.
 * @example
 * ```ts
 * const navItems = [['Home', '/', 'About', '/about',]]
 * ```
 */
export type NavItem = {
	linkText: string;
	href: string;
};

export type NavHeaderProps = Readonly<{
	/**
	 * A logo component to render in the header.
	 * @remarks should be wrapped in a link to the homepage or other relevant page.
	 * @example
	 * ```tsx
	 * export const Logo = ({ exampleLogo }: { exampleLogo: string }) => {
	 * 	// use the link and image component from your framework if applicable.
	 * 	return (
	 * 		<a href="/">
	 * 			<img src={exampleLogo} width="48" height="48" alt="Example Logo" />
	 * 		</a>
	 * 	);
	 * };
	 * ```
	 *
	 */
	Logo:
		| { src: string; alt: string; href: string; styles?: string }
		| React.ComponentType
		| React.ElementType;
	/**
	 * Items to render for the main navigation. Can be a tuple of label and href or a ReactNode.
	 */
	mainNavItems?: NavItem[] | React.ReactNode;
	/**
	 * Secondary navigation menu. Can be a tuple of label and href or a ReactNode
	 */
	secondaryNavItems?: NavItem[] | React.ReactNode;
	/**
	 * Styles to be applied to the mobile nav overlay.
	 */
	overlayStyles?: string;
	/**
	 * Styles to be applied to the omnipresent navbar.
	 */
	navbarStyles?: string;
	/**
	 * Options to pass to the FocusTrap component used to trap focus within the mobile nav overlay.
	 */
	focusTrapOptions?: FocusTrapProps['focusTrapOptions'];
	/**
	 * If a custom link component is required, like for metaframework, pass it here, otherwise an anchor tag will be used.
	 */
	linkComponent?: React.ElementType;

	/**
	 * State and setter for the open state of the mobile nav overlay.
	 * @example
	 * ```tsx
	 * const [isOpen, setIsOpen] = useState(false);
	 * const handleOpen = () => setIsOpen((prev) => !prev);
	 *
	 * <NavHeader
	 * 	mobileNavHandler={[isOpen, handleOpen]}
	 * 	...
	 * />
	 * ```
	 */
	mobileNavHandler: [boolean, () => void];
}>;

/**
 * Type guard for NavItem
 * @param arg the argument to check
 * @returns true if the arg is a {@link NavItem}
 */
export const isNavItemArray = (arg: unknown): arg is NavItem[] => {
	if (arg instanceof Array) {
		return arg
			.map((item) => {
				return (
					typeof item === 'object' &&
					item !== null &&
					'linkText' in item &&
					'href' in item
				);
			})
			.reduce((acc, curr) => acc && curr, true);
	}
	return false;
};
