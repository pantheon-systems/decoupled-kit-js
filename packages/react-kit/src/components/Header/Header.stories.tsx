import PantheonLogo from '@assets/pantheon-fist-blk.svg';
import { Button } from '@components/Button';
import { Header as NavHeader } from '@components/Header';
import { Row } from '@components/Row';
import type { Meta, StoryObj } from '@storybook/react';
import { useState } from 'react';
import { type NavHeaderProps } from './props';

const Header = ({
	mainNavItems,
	secondaryNavItems,
	Logo,
	linkComponent,
}: Pick<
	NavHeaderProps,
	| 'mainNavItems'
	| 'secondaryNavItems'
	| 'Logo'
	| 'linkComponent'
	| 'mobileNavHandler'
> & {
	manyItems: boolean;
}) => {
	const [isOpen, setIsOpen] = useState<boolean>(false);
	const handleOpen = () => setIsOpen((prev) => !prev);
	return (
		<Row type="flex" className="rk-mx-auto rk-bg-white">
			<NavHeader
				mobileNavHandler={[isOpen, handleOpen]}
				linkComponent={linkComponent}
				overlayStyles="rk-bg-white"
				Logo={Logo}
				mainNavItems={mainNavItems}
				secondaryNavItems={secondaryNavItems}
			/>
		</Row>
	);
};

const MainNavItems = ({ manyItems }: { manyItems?: boolean }) => {
	const navItems = [
		{ linkText: 'Home', href: '/' },
		{ linkText: 'Articles', href: '/articles' },
		{ linkText: 'Pages', href: '/pages' },
	];
	if (manyItems) {
		navItems.push(
			{ linkText: 'About', href: '/about' },
			{ linkText: 'Services', href: '/services' },
			{ linkText: 'Contact', href: '/contact' },
			{ linkText: 'Blog', href: '/blog' },
			{ linkText: 'Portfolio', href: '/portfolio' },
			{ linkText: 'FAQ', href: '/faq' },
			{ linkText: 'Shop', href: '/shop' },
			{ linkText: 'Events', href: '/events' },
			{ linkText: 'Gallery', href: '/gallery' },
			{ linkText: 'Team', href: '/team' },
			{ linkText: 'Testimonials', href: '/testimonials' },
			{ linkText: 'Career', href: '/career' },
			{ linkText: 'Terms of Service', href: '/terms' },
			{ linkText: 'Privacy Policy', href: '/privacy' },
			{ linkText: 'Support', href: '/support' },
			{ linkText: 'Login', href: '/login' },
			{ linkText: 'Register', href: '/register' },
			{ linkText: 'Dashboard', href: '/dashboard' },
			{ linkText: 'Logout', href: '/logout' },
		);
	}

	return navItems;
};

const NoFollowLink = ({
	children,
	className,
}: {
	children: React.ReactNode;
	className?: string;
}) => {
	return (
		<a rel="nofollow" className={className} href="javascript:void(0)">
			{children}
		</a>
	);
};
const SecondaryNavItems = () => {
	return (
		<>
			<li
				className="rk-mb-8 rk-mr-auto lg:rk-mb-0 lg:rk-ml-auto lg:rk-mr-0"
				key="docs"
			>
				<Button
					Element={NoFollowLink}
					href="https://decoupledkit.pantheon.io"
					type="secondary"
				>
					Docs
				</Button>
			</li>
			<li className="rk-mr-auto lg:rk-mx-3 lg:rk-mr-0" key="examples">
				<Button Element={NoFollowLink} href="/examples">
					Examples
				</Button>
			</li>
		</>
	);
};

const meta: Meta<typeof Header> = {
	title: 'Header/Header',
	component: Header,
	parameters: {
		layout: 'fullscreen',
	},
	args: {
		Logo: { src: PantheonLogo, alt: 'Pantheon Systems', href: '/' },
		mainNavItems: MainNavItems({ manyItems: false }),
		secondaryNavItems: <SecondaryNavItems />,
		linkComponent: NoFollowLink,
	},
	argTypes: {
		mainNavItems: {
			control: {
				type: 'radio',
			},
			options: ['Default', 'Many'],
			mapping: {
				Default: MainNavItems({ manyItems: false }),
				Many: MainNavItems({ manyItems: true }),
			},
		},
		secondaryNavItems: {
			table: {
				disable: true,
			},
		},
	},
	tags: ['autodocs'],
};

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {};
