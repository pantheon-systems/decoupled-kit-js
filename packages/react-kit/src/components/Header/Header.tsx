import { Button, IconButton } from '@components/Button';
import clsx from 'clsx';
import FocusTrap from 'focus-trap-react';
import React, { useEffect } from 'react';
import { CloseSVG, HamburgerMenuSVG } from './HeaderIcons';
import { NavHeaderProps, isNavItemArray } from './props';
/**
 * @see {@link https://live-storybook-react-kit.appa.pantheon.site/?path=/docs/header-header--docs}
 */
export const Header = ({
	Logo,
	mainNavItems,
	secondaryNavItems,
	overlayStyles,
	navbarStyles,
	focusTrapOptions,
	linkComponent,
	mobileNavHandler: [isOpen, handleOpen],
}: NavHeaderProps) => {
	const LinkComponent = linkComponent ? linkComponent : 'a';

	useEffect(() => {
		const handleKeyUp = (e: KeyboardEvent) => {
			if (e.key === 'Escape' || e.key === 'Esc') {
				handleOpen();
			}
		};
		const handleResize = () => {
			if (window.innerWidth >= 1024 && isOpen) {
				handleOpen();
			}
		};
		window.addEventListener('keyup', handleKeyUp);
		window.addEventListener('resize', handleResize);

		return () => {
			window.removeEventListener('keyup', handleKeyUp);
			window.removeEventListener('resize', handleResize);
		};
	}, [isOpen, handleOpen]);

	/**
	 * Need to use a ref here so the FocusTrap can work properly.
	 */
	const NavList = React.forwardRef<
		React.ElementRef<'nav'>,
		Pick<NavHeaderProps, 'mainNavItems' | 'secondaryNavItems'>
	>(({ mainNavItems, secondaryNavItems }, ref) => {
		const NAVLIST_STYLES = {
			nav: clsx('rk-flex rk-items-center'),
			ul: clsx(
				'lg:rk-menu-horizontal',
				isOpen
					? 'rk-menu-vertical rk-w-full lg:rk-hidden'
					: 'rk-hidden rk-flex-wrap lg:rk-flex',
			),
			overlay: clsx(
				'rk-w-full',
				isOpen
					? 'rk-absolute rk-bottom-0 rk-left-0 rk-z-10 rk-h-[calc(100%_-_72px)] rk-flex-col rk-overflow-y-auto rk-overflow-x-hidden rk-px-4 rk-pb-4 rk-pt-8 sm:rk-px-6 lg:rk-px-12 lg:rk-hidden'
					: 'rk-hidden rk-flex-wrap lg:rk-flex',
			),
		};

		const HeaderLogo = () => {
			if (Logo && React.isValidElement(Logo)) {
				return Logo;
			} else if (
				typeof Logo === 'object' &&
				'src' in Logo &&
				'alt' in Logo &&
				'href' in Logo
			) {
				const { src, alt, href } = Logo;
				return (
					<LinkComponent
						href={href}
						className={Logo?.styles || 'rk-h-10 rk-w-10'}
					>
						<img src={src} height="40" alt={alt} />
					</LinkComponent>
				);
			}
			return null;
		};

		const PrimaryNav = () => {
			if (isNavItemArray(mainNavItems)) {
				return mainNavItems?.map(({ linkText, href }) => (
					<li
						className="rk-mb-8 rk-w-full rk-justify-start rk-text-lg rk-text-black lg:rk-mx-4 lg:rk-mb-0 lg:rk-w-fit"
						key={linkText}
					>
						<LinkComponent
							className={clsx(
								'rk-link-hover rk-w-full rk-text-left',
								'rk-flex rk-w-full rk-min-w-full sm:rk-w-fit',
								typeof window !== 'undefined' &&
									window?.location?.pathname === href &&
									'rk-font-bold',
							)}
							href={href}
						>
							{linkText}
						</LinkComponent>
					</li>
				));
			} else {
				return mainNavItems;
			}
		};

		const SecondaryNav = () => {
			if (isNavItemArray(secondaryNavItems)) {
				return secondaryNavItems?.map(({ linkText, href }) => (
					<li
						className="rk-mr-auto first:rk-mb-3 lg:rk-mx-3 lg:rk-mr-0 first:lg:rk-mb-0"
						key={linkText}
					>
						<Button asChild href={href}>
							<LinkComponent>{linkText}</LinkComponent>
						</Button>
					</li>
				));
			} else {
				return secondaryNavItems;
			}
		};

		return (
			<header
				ref={ref}
				className={clsx(
					navbarStyles,
					'rk-z-10 rk-flex rk-w-full rk-items-center rk-py-4',
				)}
			>
				<HeaderLogo />
				<div className={clsx(overlayStyles, NAVLIST_STYLES.overlay)}>
					<nav
						className={clsx(NAVLIST_STYLES.nav, 'rk-mr-auto lg:rk-w-3/4')}
						aria-label="Main"
					>
						<ul className={NAVLIST_STYLES.ul}>
							<PrimaryNav />
						</ul>
					</nav>
					<hr
						className="divide-x rk-mb-8 rk-ml-auto rk-flex rk-w-full rk-min-w-full lg:rk-hidden"
						key="hr"
					/>
					<nav
						className={clsx(
							NAVLIST_STYLES.nav,
							'lg:rk-w-1/4 lg:rk-justify-end',
						)}
						aria-label="Secondary"
					>
						<ul className={NAVLIST_STYLES.ul}>
							<SecondaryNav />
						</ul>
					</nav>
				</div>
				{isOpen ? (
					<IconButton
						aria-hidden={!isOpen}
						onClick={handleOpen}
						className="rk-ml-auto lg:rk-hidden"
						data-testid="close-nav"
					>
						<CloseSVG />
					</IconButton>
				) : (
					<IconButton
						aria-hidden={isOpen}
						onClick={handleOpen}
						className="rk-ml-auto lg:rk-hidden"
						data-testid="open-nav"
					>
						<HamburgerMenuSVG />
					</IconButton>
				)}
			</header>
		);
	});

	NavList.displayName = 'NavList';

	return (
		<FocusTrap
			active={isOpen}
			paused={!isOpen}
			focusTrapOptions={focusTrapOptions}
		>
			<div className="rk-w-full">
				<NavList
					mainNavItems={mainNavItems}
					secondaryNavItems={secondaryNavItems}
				/>
			</div>
		</FocusTrap>
	);
};
