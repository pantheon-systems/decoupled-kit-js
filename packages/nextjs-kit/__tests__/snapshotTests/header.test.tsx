import { render } from '@testing-library/react';
import Link from 'next/link';
import { Header } from '../../src/components/Header/index';
/**
 * @vitest-environment jsdom
 */

vi.mock('next/router', () => ({
	useRouter: () => ({
		events: {
			on: vi.fn(),
			off: vi.fn(),
		},
	}),
}));

const navItems = [
	{
		linkText: 'Home',
		href: '/',
	},
	{
		linkText: 'Posts',
		href: '/posts',
	},
	{
		linkText: 'Pages',
		href: '/pages',
	},
];

describe('<Header />', () => {
	it("should render 'header'", () => {
		const { asFragment } = render(
			<Header
				Logo={
					<Link href="/">
						<img src={'/test-img-src'} alt="PantheonLogo" />
					</Link>
				}
				mainNavItems={navItems}
			/>,
		);

		expect(asFragment()).toMatchSnapshot();
	});
});
