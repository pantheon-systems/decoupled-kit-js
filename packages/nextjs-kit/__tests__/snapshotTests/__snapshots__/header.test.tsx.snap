// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Header /> > should render 'header' 1`] = `
<DocumentFragment>
  <div
    class="rk-w-full"
  >
    <header
      class="rk-z-10 rk-flex rk-w-full rk-items-center rk-py-4"
    >
      <a
        href="/"
      >
        <img
          alt="PantheonLogo"
          src="/test-img-src"
        />
      </a>
      <div
        class="rk-w-full rk-hidden rk-flex-wrap lg:rk-flex"
      >
        <nav
          aria-label="Main"
          class="rk-flex rk-items-center rk-mr-auto lg:rk-w-3/4"
        >
          <ul
            class="lg:rk-menu-horizontal rk-hidden rk-flex-wrap lg:rk-flex"
          >
            <li
              class="rk-mb-8 rk-w-full rk-justify-start rk-text-lg rk-text-black lg:rk-mx-4 lg:rk-mb-0 lg:rk-w-fit"
            >
              <a
                class="rk-link-hover rk-w-full rk-text-left rk-flex rk-w-full rk-min-w-full sm:rk-w-fit rk-font-bold"
                href="/"
              >
                Home
              </a>
            </li>
            <li
              class="rk-mb-8 rk-w-full rk-justify-start rk-text-lg rk-text-black lg:rk-mx-4 lg:rk-mb-0 lg:rk-w-fit"
            >
              <a
                class="rk-link-hover rk-w-full rk-text-left rk-flex rk-w-full rk-min-w-full sm:rk-w-fit"
                href="/posts"
              >
                Posts
              </a>
            </li>
            <li
              class="rk-mb-8 rk-w-full rk-justify-start rk-text-lg rk-text-black lg:rk-mx-4 lg:rk-mb-0 lg:rk-w-fit"
            >
              <a
                class="rk-link-hover rk-w-full rk-text-left rk-flex rk-w-full rk-min-w-full sm:rk-w-fit"
                href="/pages"
              >
                Pages
              </a>
            </li>
          </ul>
        </nav>
        <hr
          class="divide-x rk-mb-8 rk-ml-auto rk-flex rk-w-full rk-min-w-full lg:rk-hidden"
        />
        <nav
          aria-label="Secondary"
          class="rk-flex rk-items-center lg:rk-w-1/4 lg:rk-justify-end"
        >
          <ul
            class="lg:rk-menu-horizontal rk-hidden rk-flex-wrap lg:rk-flex"
          />
        </nav>
      </div>
      <button
        aria-hidden="false"
        class="rk-btn rk-h-10 rk-max-h-10 rk-min-h-fit rk-rounded-full rk-border-none rk-bg-white rk-p-2 rk-text-black	hover:rk-bg-neutral-100 hover:rk-text-neutral-900 active:rk-bg-neutral-300 [&>*:first-child]:rk-w-6 rk-ml-auto lg:rk-hidden"
        data-testid="open-nav"
      >
        <svg
          aria-label="open"
          fill="none"
          height="24"
          viewBox="0 0 40 40"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="20"
            cy="20"
            fill="none"
            r="20"
          />
          <path
            d="M9 12H31"
            stroke="#171717"
            stroke-linecap="round"
            stroke-width="2"
          />
          <path
            d="M9 20H31"
            stroke="#171717"
            stroke-linecap="round"
            stroke-width="2"
          />
          <path
            d="M9 28L31 28"
            stroke="#171717"
            stroke-linecap="round"
            stroke-width="2"
          />
        </svg>
      </button>
    </header>
  </div>
</DocumentFragment>
`;
